check invariant metrics after e2e tests

BenTheElder opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

Implements https://kep.k8s.io/5514

Introduces special tests which run after the entire test suite to validate invariant metrics.

Specifically starting with the declarative validation metric which compares legacy validation code and declarative validation, a non-zero value indicating a bug in api-server.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Introduces e2e tests that check component invariant metrics across the entire suite run.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://kep.k8s.io/5514
```

BenTheElder committed:
check invariant metrics after e2e tests

aojea commented on test/e2e/e2e_invariants.go:
```diff
@@ -0,0 +1,90 @@
+	}
+	apiserverMetrics, err := grabber.GrabFromAPIServer(ctx)
+	if err != nil {
+		framework.Failf("error grabbing api-server metrics: %v", err)
```
should we fail if we do not get the invariants or just log ?

BenTheElder replied:
We actually already have default tests that require fetching metrics with the grabber including apiserver specifically and they seem to be reliable, it is a very simple API call.

I think it is better to fail and detect this unless it is proven to be unreliable. In which case we can either switch to logging or mark the test as flaky until it is improved.

But I don't see any reason for this to fail other than an unhealthy API server.

If you mean should we fail if the specific metrics are missing in the response, I think so as well, that would imply a bug.

BenTheElder commented on test/e2e/e2e_invariants.go:
```diff
@@ -0,0 +1,90 @@
+	for _, sample := range samples {
+		// TODO: this logic is intentionally inverted to showcase the results of a failure
+		// It should be > 0 instead
+		if sample.Value <= 0 {
```
(This is also a dummy currently for testing purposes per the comment, will follow up on this particular line)

BenTheElder commented:
So this intentionally has the wrong logic for the metric value currently to ensure that it will fail and surface the results in CI:

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133394/pull-kubernetes-e2e-kind/1953004021803388928

<img width="1610" height="358" alt="Screenshot 2025-08-06 at 1 32 03 AM" src="https://github.com/user-attachments/assets/36e69260-7e6e-49e0-9956-ec154021f5a9" />

We can also see that it does NOT fail in the jobs specifically selecting conformance, because the "test" that enables checking the metrics at the end is skipped and checking the metrics is therefore skipped:

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133394/pull-kubernetes-conformance-kind-ga-only-parallel/1953004022029881344

When all the tests have completed, and we're otherwise happy with the shape, I'll fix the actual metric check line to be correct (there is a TODO line here https://github.com/kubernetes/kubernetes/pull/133394#discussion_r2255835587)

BenTheElder commented:
[Invoking additional optional jobs for clarity]

/test pull-kubernetes-conformance-kind-ga-only
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-e2e-kind-alpha-beta-features

aojea commented on test/e2e/OWNERS:
```diff
@@ -0,0 +1,16 @@
+  # special checks that run after the entire suite
+  # must be specially reviewed
+  "e2e_invariants.go":
+    approvers:
```
Does this guarantee that top level approver are not able to approve?

BenTheElder replied:
No, but this config guarantees auto assignment will go to this group, and there are comments throughout that you must ask.

if we want to enforce that we can refactor a bit to put the key details in a subdirectory and use no_parent_owners.

We already have no_parent_owners at the test/ level but there are a lot of people listed there.
Anyone there could already approve arbitrary changes to the test binary to add something like this, so that level of paranoia seemed unnecessary.

pohly replied:
Let's use a sub-directory, with or without no_parent. The way it is now unit tests for the code get linked into e2e.test. A separate package is just cleaner.

pohly replied:
See https://github.com/kubernetes/kubernetes/pull/133844 for my fork of this PR such that it checks for data races in log output instead of metrics.

aojea commented on test/e2e/e2e_invariants.go:
```diff
@@ -0,0 +1,105 @@
+	// this allows us to run it by default in most jobs, but it can be opted-out,
+	// does not run when selecting Conformance, and it can be tagged Flaky
+	// if we encounter issues with it
+	ginkgo.It(invariantsLeafText, func() {})
```
Should we also use label for filters?

BenTheElder commented:
The verify failures required adjusting to detect dry-run mode and skip the invariant metrics check in that case.

That's fixed now.

pohly commented on test/e2e/e2e_invariants.go:
```diff
@@ -0,0 +1,109 @@
+		return
+	}
+	// otherwise actually check invariants now
+	checkInvariants(context.TODO())
```
```suggestion
	checkInvariants(ctx)
```

And change `func(report ginkgo.Report)` to `func(ctx context.Context, report ginkgo.Report)`.

BenTheElder replied:
done.

pohly commented on test/e2e/suites.go:
```diff
@@ -35,6 +35,7 @@ func AfterSuiteActions(ctx context.Context) {
 	if framework.TestContext.ReportDir != "" {
 		framework.CoreDump(framework.TestContext.ReportDir)
 	}
+	// TODO: nothing is using this? what should we do with it?
```
Deprecate the command line option for it in 1.35, then remove it in 1.36?

pohly replied:
Or start using it, at least in some jobs? We recently had a case where it would have been nice to get more insights into control plane load during testing: https://github.com/kubernetes/kubernetes/issues/133531#issuecomment-3221083935

cc @tkashem

pohly replied:
We would need to figure out how what to do with the data, though. A simple howto somewhere would be useful.

pohly reviewed (commented):
The approach makes sense to me. One nit.

I am going to try something similar with post-run log checking.

pohly commented on test/e2e/OWNERS:
```diff
@@ -0,0 +1,16 @@
+filters:
+  # special checks that run after the entire suite
+  # must be specially reviewed
+  "e2e_invariants.go":
```
Do we want all invariant checks in a single file? Can we use a wildcard here like `e2e_*_invariants.go` and then do `e2e_invariants.go` (common definition), `e2e_metric_invariants.go` (your code) and `e2e_log_invariants.go` (something that I intent to add, for example to check for `DATA RACE` - https://github.com/kubernetes/kubernetes/pull/133834)?

pohly replied:
regexp works (https://www.kubernetes.dev/docs/guide/owners/#filters ).

BenTheElder replied:
Let's use a package.

BenTheElder replied:
moved to a pacakge.

BenTheElder commented:
Also explicit
/hold
For https://github.com/kubernetes/enhancements/pull/5514
In addition to the "WIP"

BenTheElder commented:
[bad rebase, fixed]
/remove-area code-generation
/remove-sig auth apps
/remove-kind api-change

BenTheElder commented on test/e2e/invariants/metrics.go:
```diff
@@ -0,0 +1,109 @@
+	samples, ok := metrics[metric]
+	if !ok || len(samples) == 0 {
+		framework.Failf(
+			"Invariant failed for missing metric: %v\nIf this failed on a PR, check your PR for bugs, if this failed on a periodic test please file a bug and assign the owners.\nOwners for this metric: %v",
```
We plan to bikeshed this language a bit more, per discussion on the KEP. I haven't updated it just yet in case we continue to discuss there ...

BenTheElder commented:
I updated the PR based on the KEP (including listing SIG and updating the language in the failure messages), and also added a lightweight abstraction to more readily support additional metrics later.

BenTheElder commented on test/e2e/invariants/metrics.go:
```diff
@@ -0,0 +1,162 @@
+	{
+		Metric: "apiserver_validation_declarative_validation_mismatch_total",
+		SIG:    "api-machinery",
+		Owners: []string{"aaron-prindle", "jpbetz", "thockin"},
```
FYI @aaron-prindle @jpbetz @thockin

BenTheElder replied:
ACK, I also expect we will see more metrics for other features added. Let me take a look at these.

With the new broken-down-by-identifier versions being alpha, I'm not sure if we should switch yet. Is there a thread discussing the timeline for these? Will `declarative_validation_panic_total` also be replaced by `declarative_validation_panics_total`?

lalitc375 replied:
declarative_validation_panic_total and declarative_validation_panics_total, both of these metrics are temporary and will be deprecated once the migration is complete. The `declarative_validation_panic_total` metric is a counter that will publish even in the absence of panics. The `declarative_validation_panics_total` metric is a counter vector that will publish a breakdown if any panics occur.

lalitc375 replied:
Just saw the Ben's last comment. TODO sounds good.

BenTheElder replied:
Added the panic metric and TODOs

aojea commented on test/e2e/invariants/metrics.go:
```diff
@@ -0,0 +1,162 @@
+	// this test is a sentinel for selecting the report after suite logic
+	//
+	// this allows us to run it by default in most jobs, but it can be opted-out,
+	// does not run when selecting Conformance, and it can be tagged Flaky
```
how is this implemented? how is autoskipped on conformance?

BenTheElder replied:
The sentinel test is not tagged as [Conformance] so it will not be selected when running conformance.
When the sentinel test does not run, we do not collect metrics and check the invariants in the reporter after the suite.

BenTheElder replied:
By the same logic, we can for example, tag the sentinel test as [Flaky], [Slow], etc if we needed.

The logic to check if the test ran is very simple, quick, and happens before any real check.

BenTheElder commented on test/e2e/invariants/metrics.go:
```diff
@@ -0,0 +1,162 @@
+	invariantsSelected := false
+	for _, spec := range report.SpecReports {
+		if spec.LeafNodeText == invariantMetricsLeafText {
+			invariantsSelected = spec.State.Is(ginkgotypes.SpecStatePassed)
```
@aojea this is where we decide if we will fetch the metrics and validate them or not

BenTheElder commented:
`k8s.io/kubernetes/pkg/controller: podautoscaler` https://github.com/kubernetes/kubernetes/issues/134299
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133394/pull-kubernetes-unit/1973418055530713088

BenTheElder commented:
/retest
GCE also had an unrelated flake https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133394/pull-kubernetes-e2e-gce/1973418054880595968

aojea commented on test/e2e/invariants/metrics.go:
```diff
@@ -0,0 +1,162 @@
+
+// Please speak to SIG-Testing leads before adding anything here.
+// All fields must be specified
+var apiServerMetricInvariants = []apiServerMetricInvariant{
```
can we add an unit test to validate this?

```go
package invariants

import (
	"testing"
)

func TestApiServerMetricInvariantsFieldsAreSet(t *testing.T) {
	for i, inv := range apiServerMetricInvariants {
		if inv.Metric == "" {
			t.Errorf("apiServerMetricInvariants[%d].Metric is not set", i)
		}
		if inv.SIG == "" {
			t.Errorf("apiServerMetricInvariants[%d].SIG is not set", i)
		}
		if len(inv.Owners) == 0 {
			t.Errorf("apiServerMetricInvariants[%d].Owners is not set", i)
		}
		if inv.IsValid == nil {
			t.Errorf("apiServerMetricInvariants[%d].IsValid is not set", i)
		}
	}
}
```

aojea commented:
just one comment https://github.com/kubernetes/kubernetes/pull/133394/files#r2397937663 to make automation simplify reviewers life and then LGTM

BenTheElder committed:
add test to verify invariants are fully specified

BenTheElder committed:
add apiserver_validation_declarative_validation_panic_total invariant

BenTheElder committed:
add TODO to migrate when new metrics mature and old metrics are deprecated

BenTheElder commented:
rebased, added the panic metric, added TODOs, added unit test to verify that registered apiserver metric invariants set all fields

aojea commented:
/lgtm
/approve

ship it

BenTheElder commented:
/hold cancel
